bioio_ome_zarr/writers/ome_zarr_writer.py

```diff
@@ -0,0 +1,489 @@
+                reader_set = set(reader_order)
+                writer_set = set(writer_axes)
+
+                if reader_set != writer_set:
+                    raise ValueError(
+                        "Reader axes "
+                        f"{sorted(reader_set)} do not match writer axes "
+                        f"{sorted(writer_set)}. Configure the writer axes to "
+                        "match the Reader; no implicit expansion is performed."
+                    )
+                if "t" not in reader_set:
+                    raise ValueError("Reader lacks 'T' but writer expects time.")
+
+                # ask the Reader to return data in the WRITER'S axis order
+                request_order = "".join(ax.upper() for ax in writer_axes)
+
+                # only slice axes the reader actually has
+                slice_kwargs: Dict[str, Any] = {}
+                if "t" in reader_set:
+                    slice_kwargs["T"] = slice(start_t, end_t)
+                if "c" in reader_set and channel_indexes is not None:
+                    slice_kwargs["C"] = channel_indexes
+
+                block = source.get_image_dask_data(request_order, **slice_kwargs)
+
+            else:
+                # Array path: already in writer order & ndim (including T)
+                if isinstance(source, np.ndarray):
+                    chunks = self.datasets[0].chunks
+                    assert chunks is not None, "Writer datasets must have chunks."
+                    arr = da.from_array(source, chunks=chunks)
+                else:
+                    arr = source  # dask
+
+                if arr.ndim != self.ndim:
+                    raise ValueError(
+                        "Array source ndim "
+                        f"({arr.ndim}) must match writer.ndim ({self.ndim}). "
+                        "No implicit expansion or reordering is performed."
+                    )
+                sel: List[slice] = [slice(None)] * self.ndim
+                sel[axis_t] = slice(start_t, end_t)
+                block = arr[tuple(sel)]
+
+            # Level 0: per‑T region writes
+            for k_abs in range(start_t, end_t):
+                rel = k_abs - start_t
+                sel0: List[slice] = [slice(None)] * self.ndim
+                sel0[axis_t] = slice(rel, rel + 1)
+                subset = block[tuple(sel0)]
+
+                region_one = _region_one(k_abs)
+                if self.zarr_format == 2:
+                    da.to_zarr(subset, self.datasets[0], region=region_one)
+                else:
+                    # Zarr v3: pass slice tuple directly; array objects aren't hashable
+                    da.store(
+                        subset,
+                        self.datasets[0],
+                        regions=region_one,
+                        lock=True,
+                    )
+
+            # Lower levels: resize once, then per‑T region writes
+            for lvl in range(1, self.num_levels):
```

Comment:
Feels like this and the level 0 could either be combined or use a shared function since so much of this is duplicate

Reply:
resolved in [2221420](https://github.com/bioio-devs/bioio-ome-zarr/pull/72/commits/22214202cb4c2f930f1334883bd2152284b74a83)